File: backend/src/modules/hr/hrController.js
import * as model from "./hrModel.js";

const wrap = (fn) => async (req, res) => {
  try { await fn(req, res); }
  catch (err) { console.error(err); res.status(500).json({ message: "Internal server error" }); }
};

export const listEmployees  = wrap(async (req, res) => res.json({ employees: await model.listEmployees() }));
export const getEmployee    = wrap(async (req, res) => { const e=await model.getEmployee(Number(req.params.id)); e?res.json({employee:e}):res.status(404).json({message:"Not found"}); });
export const createEmployee = wrap(async (req, res) => {
  if(!req.body.employee_code) return res.status(400).json({message:"employee_code required"});
  res.status(201).json({employee: await model.createEmployee(req.body)});
});
export const updateEmployee = wrap(async (req, res) => { const e=await model.updateEmployee(Number(req.params.id),req.body); e?res.json({employee:e}):res.status(404).json({message:"Not found"}); });

export const markAttendance = wrap(async (req, res) => {
  if(!req.body.employee_id||!req.body.date) return res.status(400).json({message:"employee_id and date required"});
  res.json({attendance: await model.markAttendance(req.body)});
});
export const getAttendance = wrap(async (req, res) => {
  const { from, to } = req.query;
  res.json({attendance: await model.getAttendance({ employee_id: Number(req.params.id), from, to })});
});
export const getAttendanceSummary = wrap(async (req, res) => {
  const { from, to } = req.query;
  if(!from||!to) return res.status(400).json({message:"from and to required"});
  res.json({summary: await model.getAttendanceSummary({ from, to })});
});

export const listPayroll      = wrap(async (req, res) => {
  const { month, year } = req.query;
  if(!month||!year) return res.status(400).json({message:"month and year required"});
  res.json({payroll: await model.listPayroll({ month, year })});
});
export const createPayroll    = wrap(async (req, res) => {
  if(!req.body.employee_id||!req.body.month||!req.body.year) return res.status(400).json({message:"employee_id, month, year required"});
  res.status(201).json({payroll: await model.createPayrollEntry(req.body)});
});
export const processPayroll   = wrap(async (req, res) => res.json({payroll: await model.processPayroll(Number(req.params.id))}));
export const payPayroll       = wrap(async (req, res) => res.json({payroll: await model.payPayroll(Number(req.params.id))}));

File: crm_mobile/lib/app/modules/hr/hr_controller.dart
import 'package:get/get.dart';

import '../../core/models/hr_models.dart';
import '../../core/network/error_utils.dart';
import '../auth/auth_controller.dart';

class HrController extends GetxController {
  final AuthController _auth = Get.find<AuthController>();

  final isLoading = false.obs;
  final isSubmitting = false.obs;
  final selectedTab = 0.obs;
  final errorMessage = ''.obs;

  final month = DateTime.now().month.obs;
  final year = DateTime.now().year.obs;
  final fromDate = ''.obs;
  final toDate = ''.obs;

  final employees = <HrEmployeeRow>[].obs;
  final attendanceSummary = <AttendanceSummaryRow>[].obs;
  final payroll = <PayrollRow>[].obs;

  @override
  void onInit() {
    super.onInit();
    final now = DateTime.now();
    fromDate.value = '${now.year}-${now.month.toString().padLeft(2, '0')}-01';
    toDate.value = '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}';
    loadAll();
  }

  Future<void> loadAll() async {
    isLoading.value = true;
    errorMessage.value = '';
    try {
      await Future.wait([loadEmployees(), loadAttendanceSummary(), loadPayroll()]);
    } catch (e) {
      errorMessage.value = userFriendlyError(e);
    } finally {
      isLoading.value = false;
    }
  }

  Future<void> loadEmployees() async {
    final res = await _auth.authorizedRequest(method: 'GET', path: '/hr/employees');
    employees.assignAll(
      (res as List).map((e) => HrEmployeeRow.fromJson(Map<String, dynamic>.from(e as Map))),
    );
  }

  Future<void> loadAttendanceSummary() async {
    final res = await _auth.authorizedRequest(
      method: 'GET',
      path: '/hr/attendance/summary?from=${fromDate.value}&to=${toDate.value}',
    );
    attendanceSummary.assignAll(
      (res as List).map((e) => AttendanceSummaryRow.fromJson(Map<String, dynamic>.from(e as Map))),
    );
  }

  Future<void> loadPayroll() async {
    final res = await _auth.authorizedRequest(
      method: 'GET',
      path: '/hr/payroll?month=${month.value}&year=${year.value}',
    );
    payroll.assignAll(
      (res as List).map((e) => PayrollRow.fromJson(Map<String, dynamic>.from(e as Map))),
    );
  }

  Future<void> markAttendance({
    required int userId,
    required String date,
    required String status,
    String? checkIn,
    String? checkOut,
  }) async {
    isSubmitting.value = true;
    try {
      await _auth.authorizedRequest(
        method: 'POST',
        path: '/hr/attendance',
        body: {
          'user_id': userId,
          'date': date,
          'status': status,
          'check_in': (checkIn ?? '').trim().isEmpty ? null : checkIn,
          'check_out': (checkOut ?? '').trim().isEmpty ? null : checkOut,
        },
      );
      await loadAttendanceSummary();
    } finally {
      isSubmitting.value = false;
    }
  }

  Future<void> createPayroll({
    required int employeeId,
    required num basic,
    num hra = 0,
    num allowances = 0,
    num deductions = 0,
    num pf = 0,
  }) async {
    isSubmitting.value = true;
    try {
      await _auth.authorizedRequest(
        method: 'POST',
        path: '/hr/payroll',
        body: {
          'employee_id': employeeId,
          'month': month.value,
          'year': year.value,
          'basic': basic,
          'hra': hra,
          'allowances': allowances,
          'deductions': deductions,
          'pf': pf,
        },
      );
      await loadPayroll();
    } finally {
      isSubmitting.value = false;
    }
  }

  Future<void> processPayroll(int payrollId) async {
    await _auth.authorizedRequest(method: 'PATCH', path: '/hr/payroll/$payrollId/process');
    await loadPayroll();
  }

  Future<void> payPayroll(int payrollId) async {
    await _auth.authorizedRequest(method: 'PATCH', path: '/hr/payroll/$payrollId/pay');
    await loadPayroll();
  }
}
